### PR TITLE
feat(adm): Periodically update suggestion data from remote settings

### LINKED ISSUE #132: Periodically update suggestion data from remote settings
In order to make sure we have the most up to date data we can, we should update the data in our remote settings suggestion provider periodically.

**Acceptance Criteria**

* Sync new data with Remote Settings periodically
* The period should be configurable
* Data that is already present should not be downloaded again
* Testing tools in merino-integration-tests should be updated

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/CONSVC-1259)

### PR DESCRIPTION
**Pull request by mythmon**
- refactor(adm): Make a separate remote settings client
- refactor(deduped-dashmap): Move deduped-map to it's own crate
- feat(adm): Periodically re-sync suggestions from Remote Settings

Note for review: `merino-adm/src/remote_settings.rs` got split up into `merino-adm/src/remote_settings/mod.rs` and `merino-adm/src/remote_settings/client.rs`. Git didn't pick on that very well, so it's a bit tough to review. Sorry about that. Most of `mod.rs` was unchanged in the first commit in the series, and in that commit `client.rs` contains most of the changes.

Fixes #132.

### CODE REVIEW

**Comment by hackebrot** [comment]
Hi @mythmon! 👋🏻 I plan to take a look at this tomorrow.

File: merino-adm/src/remote_settings/client.rs [hunk 1]
```diff
@@ -0,0 +1,609 @@
...
+}
+
+impl Record {
+    /// Get a reference to the attachment related to this record, if it exists.
+    pub fn attachment(&self) -> Option<&LazyAttachment> {
+        self.attachment.as_ref().map(Arc::as_ref)
+    }
+}
+
+impl RemoteSettingsClient {
```
Comment (pjenvey) [inline merino-adm/src/remote_settings/client.rs:225]: May I suggest moving this impl closer to the struct definition?

File: merino-adm/src/remote_settings/client.rs [hunk 2]
```diff
@@ -0,0 +1,609 @@
...
+        let mut records_url = self
+            .server_url
+            .join(&format!(
+                "v1/buckets/{}/collections/{}/changeset",
+                self.bucket_id, self.collection_id
+            ))
+            .context("Building changeset URL")
+            .map_err(SetupError::InvalidConfiguration)?;
+        records_url.query_pairs_mut().extend_pairs(&[
+            ("_expected", "0"),
```
Comment (pjenvey) [inline merino-adm/src/remote_settings/client.rs:261]: I'm having trouble finding docs for `_expected`, what does this do?

**Review by pjenvey — APPROVED** [review]

**Comment by mythmon** [comment]
Rebased to squash the `f` commits into their parents. Waiting for @hackebrot's review tomorrow to merge this.

File: merino-integration-tests/src/utils/test_tools.rs [hunk 3]
```diff
@@ -34,6 +34,9 @@ use tracing_subscriber::{fmt::MakeWriter, layer::SubscriberExt};
 ///     ).await
 /// }
 /// ```
+///
+/// # Panics
+/// May panic if tests could not be set up correctly.
```
Comment (hackebrot) [inline merino-integration-tests/src/utils/test_tools.rs:39]: How often do you expect this to happen? Does it make sense to add a troubleshooting section to the documentation for this?

**Comment by mythmon** [comment]
> I've found the implementations on the DedupedMap type in deduped-dashmap/src/lib.rs a bit hard to understand, but that may be because of my limited Rust experience.

It's a complicated thing, but don't worry it's almost entirely old code that has already been reviewed in #80. I just moved it around here.

**Comment by mythmon** [comment]
Thanks for the reviews!

**Pull request merged**